executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -58,84 +129,185 @@ impl<T: FieldElement> FinalizableData<T> {
         column_ids: &HashSet<PolyID>,
         rows: impl Iterator<Item = Row<T>>,
     ) -> Self {
-        let mut column_ids = column_ids.iter().cloned().collect::<Vec<_>>();
-        column_ids.sort();
-        let data = rows.map(Entry::InProgress).collect::<Vec<_>>();
-        Self { data, column_ids }
+        let column_ids = column_ids.iter().cloned().sorted().collect::<Vec<_>>();
+        Self {
+            pre_finalized_data: rows.collect_vec(),
+            finalized_data: CompactData::new(&column_ids),
+            post_finalized_data: Vec::new(),
+            column_ids,
+        }
     }
 
+    /// Returns the total number of rows, including non-finalized rows.
     pub fn len(&self) -> usize {
-        self.data.len()
+        self.pre_finalized_data.len() + self.finalized_data.len() + self.post_finalized_data.len()
     }
 
     pub fn is_empty(&self) -> bool {
-        self.data.is_empty()
+        self.pre_finalized_data.is_empty()
+            && self.finalized_data.is_empty()
+            && self.post_finalized_data.is_empty()
     }
 
     pub fn push(&mut self, row: Row<T>) {
-        self.data.push(Entry::InProgress(row));
+        if self.finalized_data.is_empty() && self.post_finalized_data.is_empty() {
+            self.pre_finalized_data.push(row);
+        } else {
+            self.post_finalized_data.push(row);
+        }
     }
 
     pub fn pop(&mut self) -> Option<Row<T>> {
-        match self.data.pop() {
-            Some(Entry::InProgress(row)) => Some(row),
-            Some(Entry::Finalized(_)) => panic!("Row already finalized."),
-            None => None,
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop()
+        } else if !self.finalized_data.is_empty() {
+            panic!("Row already finalized");
+        } else {
+            self.pre_finalized_data.pop()
+        }
+    }
+
+    /// Removes the last row, even if it has been finalized.
+    pub fn try_remove_last_row(&mut self) {
+        if !self.post_finalized_data.is_empty() {
+            self.post_finalized_data.pop();
+        } else if !self.finalized_data.is_empty() {
+            self.finalized_data.try_remove_last_row();
+        } else {
+            self.pre_finalized_data.pop();
         }
     }
 
     pub fn extend(&mut self, other: Self) {
-        self.data.extend(other.data);
+        if self.finalized_data.is_empty() {
+            self.pre_finalized_data
```

Comment:
can you have pre and post rows with empty finalized in the middle?

Reply:
Usually not, but I did not want to rely on that assumption.